### ISSUE TITLE
Access policy imports don't properly track include{} blocks

### ISSUE THREAD

**Issue by wolfmd** [cloudflare/terraform-provider-cloudflare#681]
In the process of importing access policies to a tfstate, I've noticed that when running 
`terraform import cloudflare_access_policy.foo <zone>/<app>/<policy>`
the resulting terraform plan is
```  
~ resource "cloudflare_access_policy" "foo" {
        application_id = <app>
        decision       = "allow"
        id             =  <id>
        name           = "allow foo"
        precedence     = 1
        zone_id        = <zone>
      + include {
          + email_domain = [
              + "foo.com",
            ]
        }
    }
```
rather than a no-op as expected. When applied, the change goes through without error and doesn't seem to break the policy. This happens regardless of the type of include used.

I've confirmed that this is the include block received from the API albeit with the modifications required by the provider to make it a list. https://www.terraform.io/docs/providers/cloudflare/r/access_group.html#email_domain
``` 
        "include": [
          {
            "email_domain": {
              "domain": "foo.com"
            }
          }
        ],
```

Possibly related, when the policy is updated in the dashboard, it gets a new UID and needs to be imported again.

**Comment by jacobbednarz** [comment]
You should be importing (with the resource structure that matches your UI resource in full) which it appears you are not in this case as the `include` is showing a diff, not a noop as you expect.

On top of that, you could be experiencing issues because you're making changes in the UI and not treating the Terraform resource as the source of truth. If you only make changes in Terraform it will handle the creation of new resources where needed.

**Comment by wolfmd** [comment]
Thanks for the response!
Unfortunately, matching the resource in full fails to compile since 
```
include {
    email_domain {
        domain = "foo"
    }
}
``` 
is invalid by the definition of the resource here: https://www.terraform.io/docs/providers/cloudflare/r/access_group.html#email_domain if it was different in some way, I would expect terraform to be reporting a change rather than a pure add. I could be missing something, though.

With regard to the changes via the UI, you're completely correct that Terraform shouuld be the source of truth, however in this case I had run into a different issue that broke an access policy for gsuite (added here: https://github.com/terraform-providers/terraform-provider-cloudflare/issues/682)

**Comment by jacobbednarz** [comment]
Your configuration isn't the correct structure; it should be

```hcl
include {
   email_domain = ["example.com"]
}
```

What comes back from the API isn't a 1:1 for the resource due to the type constraints Terraform has.

**Comment by wolfmd** [comment]
Apologies, but I'm not able to see a difference between that and what I currently have defined
```
resource "cloudflare_access_policy" "foo_allow_foo" {
  application_id = cloudflare_access_application.foo.id
  zone_id        = var.zone_id
  name           = "allow foo"
  decision       = "allow"
  precedence     = 1

  include {
    email_domain = ["foo.com"]
  }
}
```
The diff from terraform plan adds the newlines and the comma from my original comment

**Comment by jacobbednarz** [comment]
Sorry, I assumed that was your Terraform output - not your configuration as it had the delta markers. Are you able to provide a full reproduction case and we can get on the same page as your issue isn't clear here.

**Comment by wolfmd** [comment]
Sure, the best way to reproduce this would be to create an access application via the UI. Go to Access Policies -> Create Access Policy
Application Name: test
Subdomain: test
Policy name: test
Include: Emails Ending In
Domain: foo.foo

Then create a terraform file with the following contents
```
resource "cloudflare_access_application" "app_test" {
  zone_id          = <zone_id>
  name             = "test"
  domain           = "test.<domain>"
  session_duration = "24h"
}

resource "cloudflare_access_policy" "app_test_policy" {
  application_id = cloudflare_access_application.app_test.id
  zone_id        = <zone>
  name           = "test"
  decision       = "allow"
  precedence     = 1

  include {
    email_domain = ["foo.foo"]
  }
}
```
Once this is defined, find the application ID via the API (https://api.cloudflare.com/client/v4/accounts/<account>/access/apps/ for example) and the policy id using (https://api.cloudflare.com/client/v4/accounts/<account>/access/apps/<app_id>). Once that is found, you can import the resources you created like

`terraform import cloudflare_access_application.app_test <zone_id>/<application_id>`
`terraform import cloudflare_access_policy.app_test_policy <zone_id>/<application_id>/<policy_id>`

Assuming those import successfully, you can run a `terraform plan` and the results will look something like:
```
  # cloudflare_access_application.app_test will be updated in-place
  ~ resource "cloudflare_access_application" "app_test" {
        aud              = "<aud>
        domain           = "test.<domain>"
        id               = "<id>"
      + name             = "test"
        session_duration = "24h"
        zone_id          = "<zone>"
    }

  # cloudflare_access_policy.app_test_policy will be updated in-place
  ~ resource "cloudflare_access_policy" "app_test_policy" {
        application_id = "<app_id>"
        decision       = "allow"
        id             = "<policy_id>"
        name           = "test"
        precedence     = 1
        zone_id        = "<zone>"

      + include {
          + email_domain = [
              + "foo.foo",
            ]
        }
    }
```
So the plan would apply a name update to the application (which isn't great but doesn't break anything) and adds the include section which is the problem that I'm more concerned about.

Let me know if I can provide anything else to help troubleshoot! I'm not super familiar with this codebase but I was going to try to dig a bit soon

**Comment by debayande** [comment]
Can confirm; `cloudflare_access_policy` is not the only resource affected. Per my experience:

* `cloudflare_access_application` resources get imported with the `name` field set to `null`.
* `cloudflare_access_group` resources get imported with empty `include` blocks.

I'm using version `0.12.24` of Terraform and version `2.9.0` of the provider.

**Comment by zmarouf** [comment]
Same here. I just noticed this for `cloudflare_access_group`
`terraform import cloudflare_access_group.my_name <account-id>/<resource-id>` yielded `            "include": []` in my state file.

**Comment by james-callahan** [comment]
We got same issue after importing.

**Comment by jacobbednarz** [comment]
Having a look at this one, the issue is with the `Read` method and the data it gets returned. The issue is further compounded by the fact we don't err check the `d.Set` calls which ultimately led me to the problem.

```diff
--- cloudflare/resource_cloudflare_access_policy.go
+++ cloudflare/resource_cloudflare_access_policy.go
@@ -93,12 +93,16 @@ func resourceCloudflareAccessPolicyRead(d *schema.ResourceData, meta interface{}
                return fmt.Errorf("Error finding Access Policy %q: %s", d.Id(), err)
        }

        d.Set("name", accessPolicy.Name)
        d.Set("decision", accessPolicy.Decision)
        d.Set("precedence", accessPolicy.Precedence)
        d.Set("require", accessPolicy.Require)
        d.Set("exclude", accessPolicy.Exclude)
-       d.Set("include", accessPolicy.Include)
+       if err := d.Set("include", accessPolicy.Include); err != nil {
+               log.Printf("[DEBUG] oh damn! %s", err)
+       }

        return nil
 }
```

```
2021/01/11 10:53:00 [DEBUG] oh damn! include.0.email_domain: '': source data must be an array or slice, got map
```

I was scratching my head here as to why only the import was impacted and I think it's to do with `appendConditionalAccessPolicyFields` building the struct as the schema expects whereas the import just pulls it directly from the API which is slightly different types to what is expected. I'll keep digging for a solution here.

**Comment by jacobbednarz** [comment]
I've started work on this one and have a solution however it's quite verbose.

```diff
diff --git cloudflare/resource_cloudflare_access_group.go cloudflare/resource_cloudflare_access_group.go
index f7f47ae..6ba5bad 100644
--- cloudflare/resource_cloudflare_access_group.go
+++ cloudflare/resource_cloudflare_access_group.go
@@ -518,3 +521,35 @@ func BuildAccessGroupCondition(options map[string]interface{}) []interface{} {

        return group
 }
+
+func FlattenAccessGroupForSchema(accessGroup []interface{}) []map[string]interface{} {
+       data := []map[string]interface{}{}
+
+       for _, group := range accessGroup {
+               for groupKey, groupValue := range group.(map[string]interface{}) {
+
+                       switch groupKey {
+                       case "everyone":
+                               data = append(data, map[string]interface{}{
+                                       groupKey: true,
+                               })
+                       case "email_domain", "email":
+                               data = append(data, map[string]interface{}{
+                                       groupKey: extractStringSlice(groupValue.(map[string]interface{})),
+                               })
+                       }
+               }
+       }
+
+       return data
+}
+
+func extractStringSlice(a map[string]interface{}) []string {
+       strings := []string{}
+
+       for _, value := range a {
+               strings = append(strings, value.(string))
+       }
+
+       return strings
+}

diff --git cloudflare/resource_cloudflare_access_policy.go cloudflare/resource_cloudflare_access_policy.go
index 9b28a7e..c0cbe65 100644
--- cloudflare/resource_cloudflare_access_policy.go
+++ cloudflare/resource_cloudflare_access_policy.go
@@ -105,7 +105,18 @@ func resourceCloudflareAccessPolicyRead(d *schema.ResourceData, meta interface{}
                return fmt.Errorf("failed to set exclude attribute: %s", err)
        }

-       if err := d.Set("include", accessPolicy.Include); err != nil {
+       if err := d.Set("include", FlattenAccessGroupForSchema(accessPolicy.Include)); err != nil {
                return fmt.Errorf("failed to set include attribute: %s", err)
        }

```

I'll need to add conditionals for each of the keys similar to https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/cloudflare/resource_cloudflare_access_group.go#L393 in order to fix this issue.

**Comment by jacobbednarz** [comment]
this has been fixed via [v2.18.0](https://github.com/cloudflare/terraform-provider-cloudflare/releases/tag/v2.18.0)
